File: sword-means-offer-2/019/main.go
package main

import "fmt"

// @Time     : 2022/01/10 21:18
// @Author   : Ranshi
// @File     : main.go
// @Doc      : 剑指 Offer II 019. 最多删除一个字符得到回文

func validPalindrome(s string) bool {
	var check func(int, int, int) bool
	check = func(l, r, k int) bool {
		for l < r {
			if s[l] != s[r] {
				if k != 0 {
					return check(l+1, r, k-1) || check(l, r-1, k-1)
				} else {
					return false
				}
			}
			l, r = l+1, r-1
		}
		return true
	}
	return check(0, len(s)-1, 1)
}

func main() {
	fmt.Println(validPalindrome("abca"))
}

File: sword-means-offer-2/019/main.py
# -*- coding: UTF-8 -*-
# @Time     : 2022/09/12 16:52
# @Author   : Ranshi
# @File     : main.py
# @Doc      : 剑指 Offer II 019. 最多删除一个字符得到回文
class Solution:
    def validPalindrome(self, s: str) -> bool:
        def checkPalindrome(low: int, high: int):
            i, j = low, high
            while i < j:
                if s[i] != s[j]:
                    return False
                i, j = i + 1, j - 1
            return True

        low, high = 0, len(s) - 1
        while low < high:
            if s[low] == s[high]:
                low, high = low + 1, high - 1
            else:
                return checkPalindrome(low + 1, high) or checkPalindrome(low, high - 1)
        return True


if __name__ == "__main__":
    print(Solution().validPalindrome(s="aba"))
